src/BreakpointViewer.cpp

```diff
@@ -438,278 +474,365 @@ void BreakpointViewer::setBreakState()
-				QBrush color = palette().brush(runState | currentSlot ?
-					QPalette::ColorGroup::Normal : QPalette::ColorGroup::Disabled,
-					QPalette::WindowText);
-
-				QTableWidgetItem* item1 = wpTableWidget->item(wpRow, ENABLED);
-				item1->setCheckState(Qt::Checked);
-
-				auto model = wpTableWidget->model();
-				auto combo = (QComboBox*) wpTableWidget->indexWidget(model->index(wpRow, WP_TYPE));
-				combo->setCurrentIndex(static_cast<int>(bp.type) - 1);
-				int adrLen = (bp.type == Breakpoint::WATCHPOINT_IOREAD
-				           || bp.type == Breakpoint::WATCHPOINT_IOWRITE) ? 2 : 4;
-
-				QString address = hexValue(bp.address, adrLen);
-				QString regionEnd = hexValue(bp.regionEnd, adrLen);
-				QTableWidgetItem* item2 = wpTableWidget->item(wpRow, WP_REGION);
-				item2->setForeground(color);
-				item2->setText(QString("%1%2%3").arg(address)
-						.arg(regionEnd == -1 || regionEnd == address ? "" : ":")
-						.arg(regionEnd == -1 || regionEnd == address ? "" : regionEnd));
-
-				QTableWidgetItem* item3 = wpTableWidget->item(wpRow, T_CONDITION);
-				item3->setForeground(color);
-				item3->setText(bp.condition.trimmed());
-
-				QTableWidgetItem* item4 = wpTableWidget->item(wpRow, SLOT);
-				item4->setForeground(color);
-				item4->setText(QString("%1/%2")
-					.arg(bp.ps == -1 ? QChar('X') : QChar('0' + bp.ps))
-					.arg(bp.ss == -1 ? QChar('X') : QChar('0' + bp.ss)));
-
-				QTableWidgetItem* item5 = wpTableWidget->item(wpRow, SEGMENT);
-				item5->setForeground(color);
-				item5->setText(bp.segment == -1 ? "X" : QString("%1").arg(bp.segment));
-
-				QTableWidgetItem* item6 = wpTableWidget->item(wpRow, INDEX);
-				item6->setText(QString::number(index));
-
-				break;
-			}
-			case Breakpoint::CONDITION: {
-				cnRow = createTableRow(CONDITION);
-				setBreakpointStatus(CONDITION, cnRow, true);
+		BreakpointData::Type type = bp.type == Breakpoint::BREAKPOINT ? BreakpointData::BREAKPOINT
+		                         : (bp.type == Breakpoint::CONDITION  ? BreakpointData::CONDITION : BreakpointData::WATCHPOINT);
 
-				QTableWidgetItem* item1 = cnTableWidget->item(cnRow, ENABLED);
-				item1->setCheckState(Qt::Checked);
+		if (!empty[type] && (it = scanBreakpointData(bp)) != map[type].end()) {
+			it->index = index;
+			connectBreakpoint(bp.id, *it);
+			continue;
+		}
 
-				QTableWidgetItem* item2 = cnTableWidget->item(cnRow, T_CONDITION);
-				item2->setText(bp.condition.trimmed());
+		// create new BreakpointDatas
+		BreakpointData data { type, bp.id, -1, index };
+		connectBreakpoint(bp.id, data);
+	}
 
-				QTableWidgetItem* item3 = cnTableWidget->item(cnRow, INDEX);
-				item3->setText(QString::number(index));
+	// remove missing BreakpointData
+	for (auto& m : map) {
+		auto it = m.begin();
 
-				break;
+		while (it != m.end()) {
+			if (it->index == -1) {
+				auto oldit = it++;
+				m.remove(oldit->id);
+			} else {
+				it++;
 			}
-			default:
-				assert(false);
+		}
```

Comment:
Using std::map now.

Reply:
I renamed `selector` to `tables`, since there are now more things to select from (like map). `map` turned into `maps` for consistency with this change. I saw several places in the code where collections are in plural form, so I guess that's the standard.